packages/python-packages/doc-warden/warden/enforce_readme_content.py

```diff
@@ -64,20 +64,165 @@ def verify_md_readme(readme, config, section_sorting_dict):
     html_soup = bs4.BeautifulSoup(html_readme_content, "html.parser")
 
     missed_patterns = find_missed_sections(html_soup, config.required_readme_sections)
-    missed_patterns.sort(key=lambda pattern: section_sorting_dict[pattern])
 
     return (readme, missed_patterns)
 
 # within the entire readme, are there any missing sections that are expected?
 def find_missed_sections(html_soup, patterns):
-    headers = html_soup.find_all(re.compile('^h[1-2]$'))
-    missed_patterns = []
+    header_list = html_soup.find_all(re.compile('^h[1-4]$'))
+
+    flattened_patterns = flatten_pattern_config(patterns)
+    header_index = generate_header_index(header_list, flattened_patterns)
+    observed_failing_patterns = recursive_header_search(header_index, patterns, [])
+
+    return observed_failing_patterns
+
+# gets a distinct set of ALL patterns present in a config. This is
+# important because this allows us to precalculate which patterns a given header tag will match
+def flatten_pattern_config(patterns):
     observed_patterns = []
 
-    for header in headers:
-        observed_patterns.extend(match_regex_set(header.get_text(), patterns))
+    for pattern in patterns:
+        if isinstance(pattern, dict):
+            parent_pattern, child_patterns =  next(iter(pattern.items()))
```

Comment:
Assuming this is a single entry dict since we are not iterating over items.

Reply:
Yes. It will be **impossible** to make it more than single entry due to limitiations on YAML ITSELF! :D